gitlab/client.py

```diff
@@ -807,25 +814,60 @@ def http_list(
             GitlabHttpError: When the return code is not 2xx
             GitlabParsingError: If the json data could not be parsed
         """
+        # pylint: disable=too-many-return-statements
         query_data = query_data or {}
 
         # In case we want to change the default behavior at some point
         as_list = True if as_list is None else as_list
 
-        get_all = kwargs.pop("all", False)
+        get_all = kwargs.pop("all", None)
         url = self._build_url(path)
 
         page = kwargs.get("page")
 
-        if get_all is True and as_list is True:
-            return list(GitlabList(self, url, query_data, **kwargs))
+        if as_list is False:
+            # Generator requested
+            return GitlabList(self, url, query_data, **kwargs)
 
-        if page or as_list is True:
-            # pagination requested, we return a list
-            return list(GitlabList(self, url, query_data, get_next=False, **kwargs))
+        if get_all is True:
+            return list(GitlabList(self, url, query_data, **kwargs))
 
-        # No pagination, generator requested
-        return GitlabList(self, url, query_data, **kwargs)
+        # pagination requested, we return a list
+        gl_list = GitlabList(self, url, query_data, get_next=False, **kwargs)
+        items = list(gl_list)
+
+        # No warning is emitted if any of the following conditions apply:
+        # * `all=False` was set in the `list()` call.
+        # * `page` was set in the `list()` call.
+        # * GitLab did not return the `x-per-page` header.
+        # * Number of items received is less than per-page value.
+        # * Number of items received is >= total available.
+        if get_all is False:
+            return items
+        if page is not None:
+            return items
+        if gl_list.per_page is None:
+            return items
+        if len(items) < gl_list.per_page:
+            return items
+        if gl_list.total is not None and len(items) >= gl_list.total:
+            return items
```

Comment:
Since we're getting that pylint complaint, now that it's nice and flat I think this would still look ok WDYT?
```suggestion
        if (
            get_all is False
            or page is not None
            or gl_list.per_page is None
            or (len(items) < gl_list.per_page))
            or (gl_list.total is not None and len(items) >= gl_list.total)
        ):
            return items
```

Not sure if there's any easy way to check for kwargs against items in the manager's `_list_filters` to further reduce noise and address https://github.com/python-gitlab/python-gitlab/pull/1875#issuecomment-1086635196, but that would be cool to add here :)

Reply:
I removed the pylint ignore by using a nested function.